### ISSUE TITLE
Runtime "dynamic_cast error" crash when pod is present both as a direct and as a transitive dependency

### ISSUE THREAD

**Issue by Kris9000** [Adobe-Marketing-Cloud/acp-sdks#633]
Our project references the ACP pods directly, but one of our dependencies needs them too. `pod install` works fine, and the project builds as well, but as soon as we call `ACPCore.start()` our app crashes with the following error:

```
dynamic_cast error 2: One or more of the following type_info's has hidden visibility or is defined in more than one translation unit. They should all have public visibility. N20AdobeMarketingMobile23ModuleEventListenerBaseE, N20AdobeMarketingMobile35ConfigurationListenerHubSharedStateE, N20AdobeMarketingMobile15OneTimeListenerE.
```

Reproducible on both iOS 15 and tvOS 15, standalone project with just directly referenced pods works fine. Adding the frameworks directly (without CocoaPods) gives us the same issue.

**Comment by cacheung** [comment]
@Kris9000 By looking at the log, it could be you are linking the frameworks twice.  Can you double check if they are being linked from cocoapods without the manually linking?
If you can still reproduce that,  can you provide a sample app to share with us?

Best Regards,
Calise Cheung
Mobile Engineer

**Comment by Kris9000** [comment]
@cacheung Thanks for your reply. Going by `Podfile.lock` and the XCode project I can't see any double linking going on (excerpt below).
I can't really share the dependency with you as it's on a private repo, but the Podspec has the same optimistic versioning for the ACP dependencies as the main project:
```
    core.dependency 'ACPMedia', '~> 2.0'
    core.dependency 'ACPAnalytics', '~> 2.0'
```

Excerpt from main project's `Podfile.lock`:

```
PODS:
  - ACPAnalytics (2.5.1):
    - ACPAnalytics/xcframeworks (= 2.5.1)
    - ACPCore (>= 2.9.0)
  - ACPAnalytics/xcframeworks (2.5.1):
    - ACPCore (>= 2.9.0)
  - ACPCore (2.9.4):
    - ACPCore/main (= 2.9.4)
  - ACPCore/main (2.9.4)
  - ACPMedia (2.3.2):
    - ACPCore (>= 2.9.0)
    - ACPMedia/xcframeworks (= 2.3.2)
  - ACPMedia/xcframeworks (2.3.2):
    - ACPCore (>= 2.9.0)
  - Apollo (0.49.1):
    - Apollo/Core (= 0.49.1)
 
...


SPEC REPOS:
...
  https://github.com/CocoaPods/Specs.git:
    - ACPAnalytics
    - ACPCore
    - ACPMedia
...

SPEC CHECKSUMS:
  ACPAnalytics: 65bc33a03d31dc3e60f061766b629aac988d1e29
  ACPCore: 7d09a5b7233d72fa7cb4369c99a8250fd7240305
  ACPMedia: a56efa0e8d78815985f3e37960f79e33a58eb6ef
```

**Comment by cacheung** [comment]
@Kris9000 Can you also share your pod file?

Thanks,
Calise

**Comment by Kris9000** [comment]
You can find it below, [org] and [product] are my redactions, relevant pods are in `analytics_pods`, dependency with ACP dependency is `[org]Login`

```
# Make sure that the pods don't get an extra scheme.
install! 'cocoapods', deterministic_uuids: false,
                      share_schemes_for_development_pods: false,
                      generate_multiple_pod_projects: true,
                      incremental_installation: true

min_ios_version = '14.0'
min_tvos_version = '14.0'

# Git source for [org] libraries
source 'ssh://git@git.[org].be:7999/ccpod/cocoapod.git'
source 'https://github.com/CocoaPods/Specs.git'
source 'https://bitbucket.org/vualtomobile/vudrmfairplay-pods.git'

# We need to use frameworks because of Swift.
use_frameworks!

# Don't show the warnings from your Pods in Xcode.
inhibit_all_warnings!

# Set the main workspace.
workspace '[product]'

#######################
### POD definitions ###
#######################

def linter_pod
  pod 'SwiftLint', '~> 0.31'
end

def testing_pods
  pod 'Quick',            '~> 4.0'
  pod 'Nimble',           '~> 9.0'
  pod 'Nimble-Snapshots', '~> 9.2'
end

def style_dictionary_pod
   pod 'StyleDictionary', :git => 'git@bitbucket.org:[org]dev/design-system.git'
end

def sd_webimage_pod
  pod 'SDWebImageSwiftUI', '~> 2.0'
end

def resolver_pod
  pod "Resolver", '~> 1.5'
end

def apollo_pod
  pod 'Apollo', '~> 0.49.1'
end

def [org]_login_pods
  pod 'lottie-ios', '~> 3.2.3'
  pod '[org]Login', :git => "ssh://git@git.[org].be:7999/thm/ios-[org]-login.git", :commit => '4f5f6912c7a38dcf4b30f7dc59e27cbf9fd659f4'
end

def crashlytics_pod
  pod 'Firebase/Crashlytics', '~> 8.9.0'
end

def cara_pod
  pod 'Cara', :git => 'https://github.com/icapps/ios-cara.git', :branch => 'master'
end

def theoplayer_pod
  pod 'THEOplayerSDK-basic', '~> 2.9'
end

def vualto_ios_pod
  pod 'vudrmFairPlaySDK', '~> 37'
end

def vualto_tvos_pod
  pod 'vudrmFairPlaySDK-tvos', '~> 38'
end

def analytics_pods
  pod 'ACPCore', '~> 2.0'
  pod 'ACPAnalytics', '~> 2.0'
  pod 'ACPMedia', '~> 2.0'
end

##############
### Shared ###
##############

# SharedLogic

def shared_logic_pods
  baselogic_pods
  shared_ui_pods
  services_pods
end

def shared_logic_ios_pods
  shared_logic_pods
  player_ios_pods
end

def shared_logic_tvos_pods
  shared_logic_pods
  player_tvos_pods
end

# Shared UI

def shared_ui_pods
  baselogic_pods
  style_dictionary_pod
  sd_webimage_pod
  resolver_pod
end

# Services

def services_pods
  baselogic_pods
  apollo_pod
end

# BaseLogic

def baselogic_pods
  resolver_pod
  analytics_pods
end

# Player

def player_pods
  baselogic_pods
  theoplayer_pod
end

def player_ios_pods
  player_pods
  vualto_ios_pod
end

def player_tvos_pods
  player_pods
  vualto_tvos_pod
end

##############
### Mobile ###
##############

def [product]_pods
  shared_logic_ios_pods
  linter_pod
  [org]_login_pods
  crashlytics_pod
end

def storybook_pods
  linter_pod
  shared_ui_pods
end

################
### Apple TV ###
################

def [product]tv_pods
  shared_logic_tvos_pods
  cara_pod
  linter_pod
  crashlytics_pod
end

def storybooktv_pods
  linter_pod
  shared_ui_pods
end

###############
### TARGETS ###
###############

### Mobile ###

target '[product]' do
  project '[product]/[product].xcodeproj'
  platform :ios, min_ios_version

  [product]_pods

  target '[product]Tests' do
    inherit! :search_paths

    testing_pods
  end
end

target '[product]Storybook' do
  project '[product]Storybook/[product]Storybook.xcodeproj'
  platform :ios, min_ios_version

  storybook_pods
end

### Apple TV ###

target '[product]Tv' do
  project '[product]Tv/[product]Tv.xcodeproj'
  platform :tvos, min_tvos_version

  [product]tv_pods

  target '[product]TvTests' do
    inherit! :search_paths

    testing_pods
  end
end

target '[product]StorybookTV' do
  project '[product]Storybook/[product]Storybook.xcodeproj'
  platform :tvos, min_tvos_version

  storybooktv_pods
end

###############
### Shared ###
##############

abstract_target 'UI' do
  project 'Shared/[product]SharedUI/[product]SharedUI.xcodeproj'

  shared_ui_pods

  target '[product]SharedUI' do
    platform :ios, min_ios_version
  end

  target '[product]SharedUITv' do
    platform :tvos, min_tvos_version
  end

  target '[product]SharedUITests' do
    platform :ios, min_ios_version
    inherit! :search_paths

    testing_pods
  end

  target '[product]SharedUITvTests' do
    platform :tvos, min_tvos_version
    inherit! :search_paths

    testing_pods
  end

end


abstract_target 'Services' do
  project 'Shared/[product]Services/[product]Services.xcodeproj'

  services_pods

  target '[product]Services' do
    platform :ios, min_ios_version
  end

  target '[product]ServicesTv' do
    platform :tvos, min_tvos_version
  end

  target '[product]ServicesTests' do
    platform :ios, min_ios_version
    inherit! :search_paths

    testing_pods
  end

end

abstract_target 'Logic' do
  project 'Shared/[product]SharedLogic/[product]SharedLogic.xcodeproj'

  target '[product]SharedLogic' do
    platform :ios, min_ios_version

    shared_logic_ios_pods
  end

  target '[product]SharedLogicTv' do
    platform :tvos, min_tvos_version

    shared_logic_tvos_pods
  end

  target '[product]SharedLogicTests' do
    platform :ios, min_ios_version
    inherit! :search_paths

    testing_pods
  end

  target '[product]SharedLogicTvTests' do
    platform :tvos, min_tvos_version
    inherit! :search_paths

    testing_pods
  end
end

abstract_target 'Player' do
  project 'Shared/[product]Player/[product]Player.xcodeproj'

  target '[product]Player' do
    platform :ios, min_ios_version

    player_ios_pods
  end

  target '[product]PlayerTv' do
    platform :tvos, min_tvos_version

    player_tvos_pods
  end

  target '[product]PlayerTests' do
    platform :ios, min_ios_version
    inherit! :search_paths

    testing_pods
  end

end

abstract_target 'Base' do
  project 'Shared/[product]BaseLogic/[product]BaseLogic.xcodeproj'

  target '[product]BaseLogic' do
    platform :ios, min_ios_version
    inherit! :search_paths

    baselogic_pods
  end

  target '[product]BaseLogicTv' do
    platform :tvos, min_tvos_version
    inherit! :search_paths

    baselogic_pods
  end

  target '[product]BaseLogicTests' do
    platform :ios, min_ios_version
    inherit! :search_paths

    testing_pods
  end
end

####################
### Post install ###
####################

post_install do |installer|
  installer.generated_projects.each do |project|
    project.build_configurations.each do |config|
      config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = min_ios_version
      config.build_settings['TVOS_DEPLOYMENT_TARGET'] = min_tvos_version

      if config.name == 'DEV' || config.name == 'TEST'
        config.build_settings['ENABLE_TESTABILITY'] = 'YES'
      end
    end

    project.targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings['ONLY_ACTIVE_ARCH'] = 'NO'
        config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = min_ios_version
        config.build_settings['TVOS_DEPLOYMENT_TARGET'] = min_tvos_version
      end
    end
  end
end

```

**Comment by cacheung** [comment]
@Kris9000 Can you send us the full verbose log while you call ACPCore.start().  We need to look at what exactly happened in the sdk log.

Enable [ACPCore setLogLevel:ACPMobileLogLevelVerbose]; in the registerExtension section.

We will review all the info we have and get back to you.

Thanks,
Calise

**Comment by Kris9000** [comment]
Thanks in advance. Here's most of the log, lots of `duplicate symbols` logs but those are present for Apple frameworks too, so that shouldn't be the issue as long as they're the same version, which they are.

```

objc[46004]: Class ACPAnalytics is implemented in both /Users/[user]/Library/Developer/Xcode/DerivedData/[product]-blxvotrvcfmrrrafkuezrmgebvhi/Build/Products/STAG-appletvsimulator/[product]BaseLogic.framework/[product]BaseLogic (0x1114265a8) and /Users/[user]/Library/Developer/Xcode/DerivedData/[product]-blxvotrvcfmrrrafkuezrmgebvhi/Build/Products/STAG-appletvsimulator/[product]Services.framework/[product]Services (0x112cf0828). One of the two will be used. Which one is undefined.
[... a few dozen more of these ...]
objc[46004]: Class PodsDummy_ACPMedia_tvOS is implemented in both /Users/[user]/Library/Developer/Xcode/DerivedData/[product]-blxvotrvcfmrrrafkuezrmgebvhi/Build/Products/STAG-appletvsimulator/[product]BaseLogic.framework/[product]BaseLogic (0x111426df0) and /Users/[user]/Library/Developer/CoreSimulator/Devices/854626B8-3601-47A9-B8CC-AEEAD069FAF0/data/Containers/Bundle/Application/C1D0CCA6-FCFC-4AC5-AD15-52BCD9D5CE84/[product]Tv.app/[product]Tv (0x104dacca0). One of the two will be used. Which one is undefined.
2022-01-26 11:02:54.405145+0100 [product]Tv[46004:646392] [db] _LSSchemaConfigureForStore failed with error Error Domain=NSOSStatusErrorDomain Code=-10817 "([product]ll)" UserInfo={_LSFunction=_LSSchemaConfigureForStore, ExpectedSimulatorHash={length = 32, bytes = 0x15c21a50 96faa298 b8e427fd 0ff7ff3e ... a1752c17 6c8202ed }, _LSLine=405, WrongSimulatorHash={length = 32, bytes = 0x951a852d c192c8c6 15a45352 b35ca88a ... 48cb19a7 28f2aa5d }}
2022-01-26 11:02:54.406467+0100 [product]Tv[46004:646392] [db] Failed to initialize client context with error Error Domain=NSOSStatusErrorDomain Code=-10817 "([product]ll)" UserInfo={_LSFunction=_LSSchemaConfigureForStore, ExpectedSimulatorHash={length = 32, bytes = 0x15c21a50 96faa298 b8e427fd 0ff7ff3e ... a1752c17 6c8202ed }, _LSLine=405, WrongSimulatorHash={length = 32, bytes = 0x951a852d c192c8c6 15a45352 b35ca88a ... 48cb19a7 28f2aa5d }}
2022-01-26 11:02:54.409554+0100 [product]Tv[46004:646391] [db] _LSSchemaConfigureForStore failed with error Error Domain=NSOSStatusErrorDomain Code=-10817 "([product]ll)" UserInfo={_LSFunction=_LSSchemaConfigureForStore, ExpectedSimulatorHash={length = 32, bytes = 0x15c21a50 96faa298 b8e427fd 0ff7ff3e ... a1752c17 6c8202ed }, _LSLine=405, WrongSimulatorHash={length = 32, bytes = 0x951a852d c192c8c6 15a45352 b35ca88a ... 48cb19a7 28f2aa5d }}
2022-01-26 11:02:54.409699+0100 [product]Tv[46004:646391] [db] Failed to initialize client context with error Error Domain=NSOSStatusErrorDomain Code=-10817 "([product]ll)" UserInfo={_LSFunction=_LSSchemaConfigureForStore, ExpectedSimulatorHash={length = 32, bytes = 0x15c21a50 96faa298 b8e427fd 0ff7ff3e ... a1752c17 6c8202ed }, _LSLine=405, WrongSimulatorHash={length = 32, bytes = 0x951a852d c192c8c6 15a45352 b35ca88a ... 48cb19a7 28f2aa5d }}
2022-01-26 11:02:54.606292+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
2022-01-26 11:02:54.606879+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
2022-01-26 11:02:54.608863+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
2022-01-26 11:02:54.609283+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
2022-01-26 11:02:54.610612+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
2022-01-26 11:02:54.615870+0100 [product]Tv[46004:646433] dynamic_cast error 2: One or more of the following type_info's has hidden visibility or is defined in more than one translation unit. They should all have public visibility. N20AdobeMarketingMobile23ModuleEventListenerBaseE, N20AdobeMarketingMobile35ConfigurationListenerRequestContentE, N20AdobeMarketingMobile15OneTimeListenerE.
2022-01-26 11:02:54.616024+0100 [product]Tv[46004:646195] dynamic_cast error 2: One or more of the following type_info's has hidden visibility or is defined in more than one translation unit. They should all have public visibility. N20AdobeMarketingMobile6ModuleE, N20AdobeMarketingMobile13ConfigurationE, N20AdobeMarketingMobile22ModuleDetailsInterfaceE.
2022-01-26 11:02:54.616276+0100 [product]Tv[46004:646433] dynamic_cast error 2: One or more of the following type_info's has hidden visibility or is defined in more than one translation unit. They should all have public visibility. N20AdobeMarketingMobile23ModuleEventListenerBaseE, N20AdobeMarketingMobile45ConfigurationListenerLifecycleResponseContentE, N20AdobeMarketingMobile15OneTimeListenerE.
2022-01-26 11:02:54.622828+0100 [product]Tv[46004:646433] dynamic_cast error 2: One or more of the following type_info's has hidden visibility or is defined in more than one translation unit. They should all have public visibility. N20AdobeMarketingMobile23ModuleEventListenerBaseE, N20AdobeMarketingMobile35ConfigurationListenerHubSharedStateE, N20AdobeMarketingMobile15OneTimeListenerE.
2022-01-26 11:02:54.623255+0100 [product]Tv[46004:646195] [AdobeExperienceSDK TRACE <EventHub(AMSEventHub)>]: Registering extension 'Configuration' with version '2.9.4'
2022-01-26 11:02:54.624153+0100 [product]Tv[46004:646195] [AdobeExperienceSDK TRACE <Core>]: Registered Configuration module
2022-01-26 11:02:54.627810+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
2022-01-26 11:02:54.627870+0100 [product]Tv[46004:646429] [AdobeExperienceSDK TRACE <EventHub(AMSEventHub)>]: Processing Event #0 - {
    class: "Event",
    name: "Configure with AppId",
    uniqueIdentifier: "A388A60B-F3B4-4F0C-A49F-526D983E9E9C",
    source: "com.adobe.eventSource.requestContent",
    type: "com.adobe.eventType.configuration",
    pair_id: "",
    response_pair_id: "3",
    event_[product]mber: 0,
    timestamp: "1643191374625 ms",
    data: {
        "config.appId" : "[REDACTED]"
    }
}
2022-01-26 11:02:54.628180+0100 [product]Tv[46004:646195] [AdobeExperienceSDK TRACE <EventHub(AMSEventHub)>]: Registering extension 'Analytics' with version '2.5.1'
2022-01-26 11:02:54.628635+0100 [product]Tv[46004:646429] [AdobeExperienceSDK TRACE <EventHub(AMSEventHub)>]: Event #0 (com.adobe.eventType.configuration) resulted in 0 consequence events.  Time in rules was 0 milliseconds.
2022-01-26 11:02:54.629311+0100 [product]Tv[46004:646195] [AdobeExperienceSDK DEBUG <TaskExecutor>]: System info services are required to set thread name
(lldb) 
```

**Comment by Kris9000** [comment]
An update: we were able to pin this down as an issue with our dependency architecture: all modules exposed their dependencies to their client module. By directly referencing the ACP pods from our main app targets we could run Adobe Analytics normally, albeit with a less than ideal architecture where we can't have a module using ACP between our tvOS and iOS targets.

It should however be noted that only ACP ever gave us this issue and 14 other dependencies did not. The reason our dependencies work the way they do (everything linking back to the main app targets) is Apple's binary upload system: the analysis performed there doesn't go over frameworks recursively, so any indirect dependency will make an App Store upload fail. Projects would run just fine on simulators and devices from Xcode but they will never be able to be uploaded to iTunes Connect.

**Comment by cacheung** [comment]
@Kris9000 Good to know you have a workaround for this. 
We would like to understand more about your dependencies architecture, so can make sure we are looking at the right thing.
Are you able to provide us your dependency structure in a sample app?  Just we can see how you setup modules and reference them.

**Comment by Kris9000** [comment]
I'll try and link one here soon but first I have most of a sprint's worth of tracking to catch up on ;)

In any case the setup from our Podfile should be easy enough to replicate: 

- XCode: have an iOS and tvOS app target who each have their own project in the workspace, create a "Shared Logic" project in the same workspace which has both iOS and tvOS targets to which each file should be added.
- CocoaPods: have a "shared_logic_pods" target that uses the ACP pods and import that into both the shared logic *and* app targets.

The double import is admittedly unorthodox but it's an unfortunate byproduct of the static analysis performed by Apple's binary upload tool which will error on any frameworks not linked from the app target.

**Comment by praveek** [comment]
@Kris9000 
We distribute ACP SDKs as static frameworks. If the SDKs are linked more than once, it will cause undefined runtime behaviour when trying to use it.

These logs should not be ignored. 
`objc[46004]: Class ACPAnalytics is implemented in both /Users/[user]/Library/Developer/Xcode/DerivedData/[product]-blxvotrvcfmrrrafkuezrmgebvhi/Build/Products/STAG-appletvsimulator/[product]BaseLogic.framework/[product]BaseLogic (0x1114265a8) and /Users/[user]/Library/Developer/Xcode/DerivedData/[product]-blxvotrvcfmrrrafkuezrmgebvhi/Build/Products/STAG-appletvsimulator/[product]Services.framework/[product]Services (0x112cf0828). One of the two will be used. Which one is undefined.
[... a few dozen more of these ...]` 

In your case, 
1) If your app does not use ACP SDKs directly, you can remove it as dependency from the app and link it only with the target which uses the SDK. 
2) If both your app and its dependency use ACP SDKs, make your app and its dependency depend on a new target which links ACP SDKs. 

`The double import is admittedly unorthodox but it's an unfortunate byproduct of the static analysis performed by Apple's binary upload tool which will error on any frameworks not linked from the app target.`
I don't think this is true for static frameworks. Did you face this error when using ACP SDKs?

**Comment by Kris9000** [comment]
@praveek excellent point about the indirect linking only being an issue for dynamic frameworks, I got this info from a teammate and didn't test it myself. Proper static linking should have the added benefit of quicker build times too. The upload error dates back to before we started using ACP, our Podfile was restructured a while back as a workaround to it, and until we added ACP we had no issues with this approach (apart from a few duplicate symbol warnings).

Your second suggestion, linking the dependency to the app target directly, is indeed what we did. I've added "Take a long hard look at our 400 line Podfile" to our backlog and will update here with our findings ;)
